### PR TITLE
Email click tracking

### PR DESCRIPTION
**Pull request by stevehook**
## Context

We want to better measure the effectiveness of emails that we send to candidates by tracking how many result in the candidate returning to the service via links in the emails we send.

## Changes proposed in this pull request

- [x] Add an `EmailClick` model associated with `Email`.
- [x] Add a `before_action` to `CandidateInterfaceController` (base controller for the candidate interface) that checks for `utm_source` query param and creates an EmailClick record if present.
- [x] System specs to cover the end-to-end feature.
- [x] Generate the notify reference earlier in the email pipeline so that it's available when rendering emails.
- [x] Update all magic links in candidate emails to include `utm_source` query param containing the notify reference.
- [x] Add `EmailClick` model to the list of entities that gets exported to BigQuery.

## Guidance to review

- Are the tests sufficient?
- I've only tracked 'internal' links in this PR (ones that link from emails we send back to the Apply service). There are other external links that we commonly include in candidate emails, mostly to Get into teaching and Find services. We could track these if we proxied these links through Apply. Is this a worthwhile next step?
- Does it make sense to add `EmailClick`s to the list of entities that gets exported to BigQuery?

## Link to Trello card

https://trello.com/c/iUCJNzHM/4645-placeholder-make-email-tracking-production-ready

## Things to check

- [x] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: app/models/email.rb [hunk 1]
```diff
@@ -2,6 +2,7 @@ class Email < ApplicationRecord
   # Note that most, but not all, emails that we send are in regards to an
   # application form. For sign-ups and sign-ins application_form will be nil.
   belongs_to :application_form, optional: true
+  has_many :email_clicks
```
Comment (tomas-stefano) [inline app/models/email.rb:5]: Just double checking ... are the email records never destroyed? We don't need `dependent: destroy`, do we?

File: spec/system/candidate_interface/signup_and_signin/candidate_email_tracking_spec.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+  def when_i_open_the_nudge_email_and_click_on_the_link
+    email = open_email(current_candidate.email_address)
+    magic_link = email.body.match(/#{candidate_interface_sign_in_path}\/\S+/)[0]
+    expect(magic_link).to be_present
+    visit magic_link
+  end
+
+  def then_an_email_click_is_logged
+    expect(@email.reload.email_clicks).to be_present
```
Comment (tomas-stefano) [inline spec/system/candidate_interface/signup_and_signin/candidate_email_tracking_spec.rb:55]: Do you think is worth asserting the number of clicks?

**Review by tomas-stefano — APPROVED** [review]
All good 🚀. Minor questions but feel free to merge!

**Comment by stevehook** [comment]
@EmmaFrith it would be good to get your input on whether we should change the signup link. It is now significantly longer and uglier with the addition of the tracking code e.g.

image

Is it time to start hiding the URL behind a 'click here' label or something?

File: app/views/candidate_mailer/_still_interested_content.text.erb [hunk 3]
```diff
@@ -19,6 +19,4 @@ Your last application is saved. All you need to do is:
 
 <% end %>
 
-Sign in to apply again:
-
-<%= @candidate_magic_link %>
+[Sign in to apply again.](<%= @candidate_magic_link %>)
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/_still_interested_content.text.erb:22]:
```suggestion
[Sign in to apply again](<%= @candidate_magic_link %>).
```

File: app/views/candidate_mailer/application_rejected_offers_only.text.erb [hunk 4]
```diff
@@ -32,6 +32,4 @@ You’re not waiting for any other decisions.
 
 The <%= 'offer'.pluralize(@offers.length) %> will be automatically declined if you do not respond by <%= @respond_by_date %>.
 
-Sign in to your account to respond:
-
-<%= @candidate_magic_link %>
+[Sign in to your account to respond.](<%= @candidate_magic_link %>)
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/application_rejected_offers_only.text.erb:35]:
```suggestion
[Sign in to your account to respond to your offer](<%= @candidate_magic_link %>).
```

File: app/views/candidate_mailer/application_rejected_one_offer_one_awaiting_decision.text.erb [hunk 5]
```diff
@@ -24,6 +24,4 @@ You can wait until you’ve received their decision, plus 10 working days, befor
 
 You should not feel pressured to respond sooner than this.
 
-However, if you are ready to respond to your existing offer you can do so through your account:
-
-<%= @candidate_magic_link %>
+However, if you are ready to respond to your existing offer you can do so through [your account.](<%= @candidate_magic_link %>)
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/application_rejected_one_offer_one_awaiting_decision.text.erb:27]:
```suggestion
However, if you are ready, you can [respond to your existing offer through your account](<%= @candidate_magic_link %>).
```

File: app/views/candidate_mailer/application_submitted.text.erb [hunk 6]
```diff
@@ -14,9 +14,7 @@ Dear <%= @application_form.first_name %>
 
 <% end %>
 
-Sign in to your account to check the progress of your application:
-
-<%= @candidate_magic_link %>
+[Sign in to your account to check the progress of your application.](<%= @candidate_magic_link %>)
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/application_submitted.text.erb:17]:
```suggestion
[Sign in to your account to check the progress of your application](<%= @candidate_magic_link %>).
```

File: app/views/candidate_mailer/application_withdrawn_on_request_one_offer_one_awaiting_decision.text.erb [hunk 7]
```diff
@@ -18,6 +18,4 @@ You can wait until you’ve received their decision, plus 10 working days, befor
 
 You should not feel pressured to respond sooner than this.
 
-However, if you are ready to respond to your existing offer you can do so through your account:
-
-<%= @candidate_magic_link %>
+However, if you are ready to respond to your existing offer you can do so through [your account.](<%= @candidate_magic_link %>)
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/application_withdrawn_on_request_one_offer_one_awaiting_decision.text.erb:21]:
```suggestion
However, if you are ready, you can [respond to your existing offer you can do so through your account](<%= @candidate_magic_link %>).
```

File: app/views/candidate_mailer/changed_offer.text.erb [hunk 8]
```diff
@@ -34,9 +34,7 @@ If you do not respond by <%= @application_choice.decline_by_default_at.to_fs(:go
   <%= render "offer_list" %>
 <% end %>
 
-Sign in to your account to respond:
-
-<%= candidate_magic_link(@application_choice.application_form.candidate) %>
+[Sign in to your account to respond.](<%= candidate_magic_link(@application_choice.application_form.candidate) %>)
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/changed_offer.text.erb:37]:
```suggestion
[Sign in to your account to respond to your offer](<%= candidate_magic_link(@application_choice.application_form.candidate) %>).
```

File: app/views/candidate_mailer/chase_candidate_decision.text.erb [hunk 9]
```diff
@@ -16,6 +16,4 @@ If you do not respond by <%= @dbd_date %>, your offer will be declined automatic
 
 <% end %>
 
-Sign in to your account to respond:
-
-<%= candidate_magic_link(@application_form.candidate) %>
+[Sign in to your account to respond.](<%= candidate_magic_link(@application_form.candidate) %>)
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/chase_candidate_decision.text.erb:19]:
```suggestion
[Sign in to your account to respond](<%= candidate_magic_link(@application_form.candidate) %>).
```

File: app/views/candidate_mailer/chase_reference.text.erb [hunk 10]
```diff
@@ -4,8 +4,6 @@ You can add as many referees as you like to increase the chances of getting 2 re
 
 You can also send one reminder to each referee from your account.
 
-Manage your references:
-
-<%= candidate_magic_link(@application_form.candidate) %>
+[Manage your references.](<%= candidate_magic_link(@application_form.candidate) %>)
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/chase_reference.text.erb:7]:
```suggestion
[Manage your references](<%= candidate_magic_link(@application_form.candidate) %>).
```

File: app/views/candidate_mailer/eoc_deadline_reminder.erb [hunk 11]
```diff
@@ -2,9 +2,7 @@
 Dear <%= @application_form.first_name %>
 <% end %>
 
-Submit your application as soon as you can to get on a course starting in the <%= RecruitmentCycle.current_year %> to <%= RecruitmentCycle.next_year %> academic year:
-
-<%= candidate_magic_link(@application_form.candidate) %>
+[Submit your application as soon as you can to get on a course starting in the <%= RecruitmentCycle.current_year %> to <%= RecruitmentCycle.next_year %> academic year.](<%= candidate_magic_link(@application_form.candidate) %>)
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/eoc_deadline_reminder.erb:5]:
```suggestion
[Submit your application](<%= candidate_magic_link(@application_form.candidate) %>) as soon as you can to get on a course starting in the <%= RecruitmentCycle.current_year %> to <%= RecruitmentCycle.next_year %> academic year.
```

File: app/views/candidate_mailer/feedback_received_for_application_rejected_by_default.text.erb [hunk 12]
```diff
@@ -9,7 +9,5 @@ They’ve now given you the following feedback:
 <%= render 'reasons_for_rejection' %>
 
 <% if @show_apply_again_guidance %>
-  If this feedback was useful, consider using it to strengthen your application and apply again:
-
-  <%= @candidate_magic_link %>
+  If this feedback was useful, consider using it to strengthen your application and [apply again](<%= @candidate_magic_link %>).
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/feedback_received_for_application_rejected_by_default.text.erb:12]:
```suggestion
  If this was useful, [use your feedback to strengthen your application and apply again](<%= @candidate_magic_link %>).
```

File: app/views/candidate_mailer/find_another_course.text.erb [hunk 13]
```diff
@@ -12,8 +12,6 @@ Find out if there are any other suitable courses available:
 
 <%= I18n.t('find_postgraduate_teacher_training.production_url') %>
 
-If you find a course, sign in to your account to add the course as soon as possible:
-
-<%= candidate_magic_link(@candidate) %>
+If you find a course, [sign in to your account to add the course as soon as possible](<%= candidate_magic_link(@candidate) %>).
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/find_another_course.text.erb:15]: is this email still in use? haven't seen this for aaaaages

File: app/views/candidate_mailer/find_has_opened.erb [hunk 14]
```diff
@@ -10,9 +10,7 @@ Find your courses:
 
 Apply early to have the best chance of getting onto the course you want, as courses fill up throughout the year.
 
-Get your application ready:
-
-<%= candidate_magic_link(@application_form.candidate) %>
+[Get your application ready.](<%= candidate_magic_link(@application_form.candidate) %>):
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/find_has_opened.erb:13]:
```suggestion
[Get your application ready](<%= candidate_magic_link(@application_form.candidate) %>).
```

File: app/views/candidate_mailer/new_offer/multiple_offers.text.erb [hunk 15]
```diff
@@ -16,9 +16,7 @@ If you do not respond by <%= @application_choice.decline_by_default_at.to_fs(:go
 
 <%= render "offer_list" %>
 
-Sign in to your account to respond:
-
-<%= candidate_magic_link(@application_choice.application_form.candidate) %>
+[Sign in to your account to respond.](<%= candidate_magic_link(@application_choice.application_form.candidate) %>)
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/new_offer/multiple_offers.text.erb:19]:
```suggestion
[Sign in to your account to respond to your offer](<%= candidate_magic_link(@application_choice.application_form.candidate) %>).
```

File: app/views/candidate_mailer/new_referee_request.text.erb [hunk 16]
```diff
@@ -12,8 +12,6 @@ You can add as many referees as you like to increase the chances of getting 2 re
 
 You can also send one reminder to each referee from your account.
 
-Manage your references:
-
-<%= candidate_magic_link(@candidate) %>
+[Manage your references.](<%= candidate_magic_link(@candidate) %>):
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/new_referee_request.text.erb:15]:
```suggestion
[Manage your references](<%= candidate_magic_link(@candidate) %>).
```

File: app/views/candidate_mailer/nudge_unsubmitted.text.erb [hunk 17]
```diff
@@ -2,9 +2,7 @@ Dear <%= @application_form.first_name %>
 
 It looks like your teacher training application is ready to submit.
 
-If you’re ready to submit now, sign in to your account:
-
-<%= candidate_magic_link(@application_form.candidate) %>
+[If you’re ready to submit now, sign in to your account.](<%= candidate_magic_link(@application_form.candidate) %>)
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/nudge_unsubmitted.text.erb:5]:
```suggestion
If you’re ready, [sign in to your account to submit your application now](<%= candidate_magic_link(@application_form.candidate) %>).
```

File: app/views/candidate_mailer/nudge_unsubmitted_with_incomplete_courses.text.erb [hunk 18]
```diff
@@ -2,9 +2,7 @@ Dear <%= @application_form.first_name %>
 
 You’ve not added courses to your teacher training application yet.
 
-Sign in to your account to add up to 3 courses:
-
-<%= candidate_magic_link(@application_form.candidate) %>
+[Sign in to your account to add up to 3 courses.](<%= candidate_magic_link(@application_form.candidate) %>)
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/nudge_unsubmitted_with_incomplete_courses.text.erb:5]:
```suggestion
[Sign in to your account to add up to 3 courses](<%= candidate_magic_link(@application_form.candidate) %>).
```

File: app/views/candidate_mailer/nudge_unsubmitted_with_incomplete_personal_statement.text.erb [hunk 19]
```diff
@@ -2,9 +2,7 @@ Dear <%= @application_form.first_name %>
 
 You have not marked your personal statement as complete.
 
-Sign in to your account if you’re ready to submit your teacher training application:
-
-<%= candidate_magic_link(@application_form.candidate) %>
+[Sign in to your account if you’re ready to submit your teacher training application.](<%= candidate_magic_link(@application_form.candidate) %>)
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/nudge_unsubmitted_with_incomplete_personal_statement.text.erb:5]:
```suggestion
[Sign in to your account to complete your teacher training application](<%= candidate_magic_link(@application_form.candidate) %>).
```

File: app/views/candidate_mailer/find_another_course.text.erb (hunk 13 above)
Comment (stevehook) [reply app/views/candidate_mailer/find_another_course.text.erb:15]: No, it doesn't seem to be referred to from anywhere. Will delete.

File: app/views/candidate_mailer/offer_accepted.text.erb [hunk 20]
```diff
@@ -4,9 +4,7 @@ You’ve accepted <%= @provider_name %>’s offer to study <%= @course_name_and_
 
 If you meet your conditions, you’ll start <%= @course_name_and_code %> <%= @provider_name %> on <%= @start_date %>.
 
-Check your conditions by signing in to your account:
-
-<%= candidate_magic_link(@application_form.candidate) %>
+[Check your conditions by signing in to your account.](<%= candidate_magic_link(@application_form.candidate) %>)
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/offer_accepted.text.erb:7]:
```suggestion
[Sign in to your account to check your offer conditions](<%= candidate_magic_link(@application_form.candidate) %>).
```

File: app/views/candidate_mailer/reference_received.text.erb [hunk 21]
```diff
@@ -3,21 +3,20 @@ You have a reference from <%= @reference.name %>.
 <% if @selected_references.count >= ApplicationForm::REQUIRED_REFERENCE_SELECTIONS %>
 You’ve selected 2 references to submit with your application already, but you can change your selection if you want.
 
-Sign in to complete your application:
+[Sign in to complete your application.](<%= candidate_magic_link(@application_form.candidate) %>)
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/reference_received.text.erb:6]:
```suggestion
[Sign in to complete your application](<%= candidate_magic_link(@application_form.candidate) %>).
```

File: app/views/candidate_mailer/reference_received.text.erb [hunk 22]
```diff
@@ -3,21 +3,20 @@ You have a reference from <%= @reference.name %>.
...
 <% elsif @provided_references.count > ApplicationForm::REQUIRED_REFERENCE_SELECTIONS %>
 You have more than enough references to send your application to training providers.
 
-Sign in and select the 2 references you want to include in your application:
+[Sign in and select the 2 references you want to include in your application.](<%= candidate_magic_link(@application_form.candidate) %>)
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/reference_received.text.erb:10]:
```suggestion
[Sign in to select 2 references to include in your application](<%= candidate_magic_link(@application_form.candidate) %>).
```

File: app/views/candidate_mailer/reference_received.text.erb [hunk 23]
```diff
@@ -3,21 +3,20 @@ You have a reference from <%= @reference.name %>.
...
 <% elsif @provided_references.count == ApplicationForm::REQUIRED_REFERENCE_SELECTIONS %>
 You have enough references to send your application to training providers.
 
-Sign in and select the 2 references you want to include in your application:
+[Sign in and select the 2 references you want to include in your application.](<%= candidate_magic_link(@application_form.candidate) %>)
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/reference_received.text.erb:14]:
```suggestion
[Sign in to select 2 references to include in your application.](<%= candidate_magic_link(@application_form.candidate) %>)
```

File: app/views/candidate_mailer/reference_received.text.erb [hunk 24]
```diff
@@ -3,21 +3,20 @@ You have a reference from <%= @reference.name %>.
...
 <% else %>
 You need another reference before you can send your application to training providers.
 
 You can request as many references as you like to increase the chances of getting 2 quickly.
 
-Sign in to manage your reference requests:
+[Sign in to manage your reference requests.](<%= candidate_magic_link(@application_form.candidate) %>)
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/reference_received.text.erb:20]:
```suggestion
[Sign in to manage your reference requests](<%= candidate_magic_link(@application_form.candidate) %>).
```

File: app/views/candidate_mailer/reinstated_offer.text.erb [hunk 25]
```diff
@@ -13,6 +13,4 @@ The course starts in <%= @course_option.course.start_date.to_fs(:month_and_year)
   They’ll let you know if they need further information before you start training.
 <% end %>
 
-Sign in to your account to see your offer details:
-
-<%= candidate_magic_link(@application_choice.application_form.candidate) %>
+[Sign in to your account to see your offer details.](<%= candidate_magic_link(@application_choice.application_form.candidate) %>)
```
Comment (EmmaFrith) [inline app/views/candidate_mailer/reinstated_offer.text.erb:16]:
```suggestion
[Sign in to your account to see your offer details](<%= candidate_magic_link(@application_choice.application_form.candidate) %>).
```

File: app/views/candidate_mailer/find_another_course.text.erb [hunk 26]
```diff
@@ -1,19 +0,0 @@
-Dear <%= @application_form.first_name %>
```
Comment (MylesJarvis) [inline app/views/candidate_mailer/find_another_course.text.erb:1]: Why is this deleted @stevehook? Is it because we currently don't call it anywhere?

I seem to remember I drafted a ticket up about this - we probably should be calling it and letting candidates now.

What are your thoughts?
